### We import the essential Python libraries required for:
- Data manipulation (Pandas, NumPy)t
- Visualization (Plotly)

In [6]:
#Importing Necessary libraries
import pandas as pd      # for data manipulation
import numpy as np       # for numerical calculations
import plotly.express as px  # for visualizations

### 1.)Loading the Dataset:
Loading the Nassau Candy Distributor shipment dataset.
The dataset contains order and shipment records across multiple US regions.

In [104]:
# Loading the dataset from the data folder
df=pd.read_csv(r"A:\UM\nassau_candy_project\data\Nassau Candy Distributor.csv")
df.head()# Displaying first 5 rows to verify successful loading

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Country/Region,City,State/Province,Postal Code,Division,Region,Product ID,Product Name,Sales,Units,Gross Profit,Cost
0,1,US-2021-103800-CHO-MIL-31000,03-01-2024,30-06-2026,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.50,2,4.22,2.28
1,2,US-2021-112326-CHO-TRI-54000,04-01-2024,01-07-2026,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-TRI-54000,Wonka Bar - Triple Dazzle Caramel,7.50,2,4.90,2.60
2,3,US-2021-112326-CHO-NUT-13000,04-01-2024,01-07-2026,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-NUT-13000,Wonka Bar - Nutty Crunch Surprise,10.47,3,7.47,3.00
3,4,US-2021-112326-CHO-SCR-58000,04-01-2024,01-07-2026,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-SCR-58000,Wonka Bar -Scrumdiddlyumptious,10.80,3,7.50,3.30
4,5,US-2021-141817-CHO-TRI-54000,05-01-2024,05-07-2026,Standard Class,141817,United States,Philadelphia,Pennsylvania,19143,Chocolate,Atlantic,CHO-TRI-54000,Wonka Bar - Triple Dazzle Caramel,11.25,3,7.35,3.90


### 2.)Knowing the number of rows and columns ,Column names,Data Types of columns,Missing Values(if any)
it helps us to understand the scale of operations we are dealing with 
at Nassau Candy Distributor.

In [108]:
# Checking the shape of dataset- rows and columns
print("Shape:",df.shape)
# Checking all column names
print("\nColumns:\n",df.columns.tolist())
# Checking data types of each column
print("\nDataTypes:\n",df.dtypes)
# Checking missing values in the dataset
print("\nMissing Values:\n",df.isnull().sum())

Shape: (10194, 19)

Columns:
 ['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode', 'Customer ID', 'Country/Region', 'City', 'State/Province', 'Postal Code', 'Division', 'Region', 'Product ID', 'Product Name', 'Sales', 'Units', 'Gross Profit', 'Cost', 'Lead Time']

DataTypes:
 Row ID                     int64
Order ID                  object
Order Date        datetime64[ns]
Ship Date         datetime64[ns]
Ship Mode                 object
Customer ID                int64
Country/Region            object
City                      object
State/Province            object
Postal Code               object
Division                  object
Region                    object
Product ID                object
Product Name              object
Sales                    float64
Units                      int64
Gross Profit             float64
Cost                     float64
Lead Time                  int64
dtype: object

Missing Values:
 Row ID            0
Order ID          0
Order Date    

### 2.)Finding
.Our dataset Contains 10194 Rows and 19 Columns
.Column Names and DataType of Each Column is shown in Output
.There is No missing Value in Any column


## 3.) Order Date Range
Checking the earliest and latest order dates in the dataset.
This tells us the time period our analysis covers.

In [109]:
# checking the date range of all orders in our dataset
print(f"Earliest Order Date: {df['Order Date'].min()}")
print(f"Latest Order Date: {df['Order Date'].max()}") #YY-MM-DD

Earliest Order Date: 2024-01-02 00:00:00
Latest Order Date: 2025-12-31 00:00:00


### 3.)Finding:
Our dataset covers orders from '2024-01-02' to '2025-12-31' .
This represents a '1' year period of shipment data.

### 4.) Dataset Coverage Analysis
Counting the total number of unique customers,products ,
states and cities in the dataset . This tells us the scale
and geographic spread of Nassau Candy's operations.

In [119]:
# how many different customers placed orders?
print(f"Unique Customers: {df['Customer ID'].nunique()}")
# how many different products were shipped?
print(f"Unique Products: {df['Product ID'].nunique()}")
# how many states received shipments?
print(f"Unique States: {df['State/Province'].nunique()}")
# how many cities received shipments?
print(f"Unique Cities: {df['City'].nunique()}")

Unique Customers: 5044
Unique Products: 15
Unique States: 59
Unique Cities: 542


### 4.) Findings
Nassau Candy Distributor has 5044 unique customers
across 542 unique cities and 59 states in the US.
A total of 15 unique products were shipped during
the period covered in this dataset.

## 5.)- Checking for Duplicate Records
Checking if any orders have been recorded more than once.
Duplicate records can inflate route volumes and distort 
our efficiency calculations.

In [120]:
# are there any orders recorded twice by mistake?
dup = df.duplicated().sum()
print(f"There is '{dup}' Duplicate Rows")

There is '0' Duplicate Rows


### 5.)Finding:
NO duplicate records were found in the dataset.

### 6.)Date Conversion and Lead Time Calculation & Calculating key statistics for Shipping Lead Time.
 
Converting Order Date and Ship Date from object (text) to datetime format.
This allows us to perform date arithmetic to calculate Shipping Lead Time.

Formula: Lead Time = Ship Date - Order Date

In [106]:
# Converting date columns from object to datetime format
df['Order Date']=pd.to_datetime(df['Order Date'],format='%d-%m-%Y')
df['Ship Date']=pd.to_datetime(df['Ship Date'],format='%d-%m-%Y')

# Calculating shipping lead time in days
# Lead Time = number of days between order placement and shipment
df['Lead Time']=abs(df['Ship Date']-df['Order Date']).dt.days

# Displaying lead time statistics
print(f"Minimum Lead Time:{df['Lead Time'].min()} days")
print(f"Maximum Lead Time:{df['Lead Time'].max()} days")
print(f"Average Lead Time:{df['Lead Time'].mean():.2f} days")
print(f"Most Common Lead Time:{df['Lead Time'].mode()[0]} days") #Unimodal(Lead time)

Minimum Lead Time:904 days
Maximum Lead Time:1642 days
Average Lead Time:1320.84 days
Most Common Lead Time:1273 days


### 6.)Finding:
Lead Time column successfully created for all orders,Values represent the number of days between order placement and shipment for each record.
The Minimum Lead Time Observed:904 days
The Maximum Lead Time Observed :1642 days
The average shipping Lead time across all orders is 1321 days.
The maximum Lead time observed is 1642 days which indicates potential delays on certain routes.

### 7.)Identifying Invalid Lead Times

Checking for records where Lead Time is zero or negative.
These records are logically impossible - an order cannot 
be shipped on the same day or before it was placed.
Such records are considered data entry errors.

In [121]:
# catching orders that show zero or negative lead times
# these dont make sense in real life so we need to know
# how many exist before removing them
negative_lead = df[df['Lead Time'] <= 0].shape[0]
print(f"Invalid Lead Times (zero or negative): {negative_lead}")

Invalid Lead Times (zero or negative): 0


### 7.)Finding:
No records with invalid lead times were identified.
These records will be excluded from all further analysis
to maintain data integrity.

### 8.)Lead Time Distribution
Examining how lead times are distributed across all orders.
This tells us whether most orders are delivered quickly 
or if delays are spread across many orders.

In [118]:
# this shows us how many orders had each lead time value
print(df['Lead Time'].value_counts().sort_index())

Lead Time
904       89
905       40
906      289
907      176
908      591
909      519
910      207
911      134
912        4
915        2
1269     258
1270     156
1271     610
1272     456
1273    1326
1274    1035
1275     612
1276     293
1277      18
1634       3
1635     179
1636     190
1637     449
1638     411
1639     896
1640     657
1641     397
1642     197
Name: count, dtype: int64


### 8.) Finding
The lead time distribution shows that majority of orders 
are delivered within 1273  days. Orders exceeding days 
are considered delayed and will be flagged in our 
route efficiency analysis.

### 9.)Extracting Product to Factory Mapping, Mapping Products to Factories

Every product in the Nassau Candy dataset is manufactured
at a specific factory. Since the dataset does not contain
a Factory column directly, we load the product-factory
correlation from an external CSV file and map it to our
main dataset using Product Name as the key.

This step is critical because our entire route analysis
depends on knowing which factory each order originated from.

Mapping Key: Product Name → Factory Name
I will extract the unique product and division Combinations directly from the dataset itself. 

In [124]:
# first lets see what unique products and divisions exist
# in our dataset
print(df[['Product Name', 'Division']].drop_duplicates().sort_values('Division'))

                           Product Name   Division
0            Wonka Bar - Milk Chocolate  Chocolate
1     Wonka Bar - Triple Dazzle Caramel  Chocolate
2     Wonka Bar - Nutty Crunch Surprise  Chocolate
3        Wonka Bar -Scrumdiddlyumptious  Chocolate
13            Wonka Bar - Fudge Mallows  Chocolate
19                            Wonka Gum      Other
26                            Kazookles      Other
136                  Lickable Wallpaper      Other
191                Fizzy Lifting Drinks      Sugar
226                         Laffy Taffy      Sugar
973                           SweeTARTS      Sugar
1490                              Nerds      Sugar
4435                        Hair Toffee      Sugar
4729             Everlasting Gobstopper      Sugar
5007                            Fun Dip      Sugar


In [157]:
### Loading Factory Dataset
# factory mapping data is sourced from the project description
# stored as an external csv file to keep data and code separate
factory_df = pd.read_csv(r"A:\UM\nassau_candy_project/data/factory_mapping.csv")

# converting the csv into a dictionary
# key = Product Name, value = Factory Name
factory_dic = dict(zip(factory_df['Product Name'], factory_df['Factory']))

# applying the dictionary mapping to create Factory column
# every product name gets replaced with its factory name
df['Factory'] = df['Product Name'].map(factory_dic)

# verifying mapping worked correctly
print(f"Unmapped Products: {df['Factory'].isnull().sum()}")
print(df[['Product Name','Factory']].drop_duplicates())

Unmapped Products: 0
                           Product Name            Factory
0            Wonka Bar - Milk Chocolate    Wicked Choccy's
1     Wonka Bar - Triple Dazzle Caramel    Wicked Choccy's
2     Wonka Bar - Nutty Crunch Surprise      Lot's O' Nuts
3        Wonka Bar -Scrumdiddlyumptious      Lot's O' Nuts
13            Wonka Bar - Fudge Mallows      Lot's O' Nuts
19                            Wonka Gum     Secret Factory
26                            Kazookles  The Other Factory
136                  Lickable Wallpaper     Secret Factory
191                Fizzy Lifting Drinks        Sugar Shack
226                         Laffy Taffy        Sugar Shack
973                           SweeTARTS        Sugar Shack
1490                              Nerds        Sugar Shack
4435                        Hair Toffee  The Other Factory
4729             Everlasting Gobstopper     Secret Factory
5007                            Fun Dip        Sugar Shack


### 9.) Finding
All 10194  products successfully mapped to their respective factories.
Zero unmapped products confirms complete and accurate factory assignment.
Nassau Candy Distributor operates through 5 factories across the US:
- Lot's O' Nuts
- Wicked Choccy's
- Sugar Shack
- Secret Factory
- The Other Factory

## # Step 11 - Creating Route Column

A route is defined as the shipping path from a factory
to a customer state. This is the most important feature
of our entire analysis. Every efficiency insight, KPI
and bottleneck analysis will be based on this column.

Formula: Route = Factory Name → Customer State

In [159]:
# combining factory name and customer state
# to create a unique route identifier for each order
df['Route'] = df['Factory'] + ' → ' + df['State/Province']

# also creating a region level route
# for higher level geographic analysis
df['Region Route'] = df['Factory'] + ' → ' + df['Region']

# checking how many unique routes exist
print(f"Total Unique State Routes: {df['Route'].nunique()}")
print(f"Total Unique Region Routes: {df['Region Route'].nunique()}")

# checking top 10 most common routes
print("\nTop 10 Most Common Routes:")
print(df['Route'].value_counts().head(10)) #Since Value_counts() arranges in desecnding order

Total Unique State Routes: 196
Total Unique Region Routes: 20

Top 10 Most Common Routes:
Route
Lot's O' Nuts → California      1125
Wicked Choccy's → California     823
Lot's O' Nuts → New York         645
Lot's O' Nuts → Texas            569
Wicked Choccy's → New York       436
Wicked Choccy's → Texas          388
Lot's O' Nuts → Pennsylvania     331
Lot's O' Nuts → Washington       269
Lot's O' Nuts → Illinois         267
Lot's O' Nuts → Ohio             262
Name: count, dtype: int64
